KAFKA-19053: Remove FetchResponse#of which is not used in production …

apalan60 opened the pull request:
Removed the unused FetchResponse#of that is not used in production. The
test cases that originally invoked this method have been updated to call
the other
[FetchResponse#of](https://github.com/apache/kafka/blob/6af849f8642d93877e64dace272cee1551620862/clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java#L232),
which is currently used by ```KafkaApis```, to maintain the integrity of
the tests.

Reviewers: Jun Rao <junrao@gmail.com>, PoAn Yang <payang@apache.org>,
 Chia-Ping Tsai <chia7712@gmail.com>

apalan60 committed:
KAFKA-19053: Remove FetchResponse#of which is not used in production anymore

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java:
```diff
@@ -220,14 +220,6 @@ public static int recordsSize(FetchResponseData.PartitionData partition) {
         return partition.records() == null ? 0 : partition.records().sizeInBytes();
     }
 
-    // TODO: remove as a part of KAFKA-12410
```
Could we remove the TODO for the other `of` constructor?

apalan60 replied:
@junrao

Thanks for your review.
The other ```of``` constructor is currently used by ```KafkaApis``` in production code. I'm looking into alternative approaches to remove this dependency, so that we can achieve the deletion behavior mentioned in the TODO comment.  I'll update the thread if I manage to come up with a viable solution. 

If I've misunderstood anything, please let me know.

chia7712 replied:
@junrao the another `of` is still used by fetch session, and @TaiJuWu already take over the KAFKA-12410 which will refactor the code to remove another `of` static method

apalan60 replied:
Thanks @chia7712  for helping clarify this issue.

junrao reviewed (commented):
@apalan60 : Thanks for the PR. Left a comment.

ijuma commented:
Are all the tests we deleted for the `of` method? Or are these tests for other functionality and they just happened to use `of` to create the fetch response?

FrankYang0529 commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java:
```diff
@@ -3900,86 +3653,6 @@ private Set<Node> nodesRequested(List<NetworkClientDelegate.UnsentRequest> reque
-                        .setLastStableOffset(lastStableOffset)
-                        .setLogStartOffset(0)
-                        .setRecords(records));
-        return FetchResponse.of(Errors.NONE, throttleTime, sessionId, new LinkedHashMap<>(partitions));
```
The `FetchResponse.of` is not used in production code, but we still need to use another function to replace it in test code. Probably, you can modify this kind of function like following and add related `client.prepareResponse` back in the test code.

```java
    private FetchResponse fullFetchResponse(int sessionId, TopicIdPartition tp, MemoryRecords records, Errors error, long hw,
                                            long lastStableOffset, int throttleTime) {
        return new FetchResponse(new FetchResponseData()
            .setThrottleTimeMs(throttleTime)
            .setErrorCode(Errors.NONE.code())
            .setSessionId(sessionId)
            .setResponses(List.of(
                new FetchResponseData.FetchableTopicResponse()
                    .setTopic(tp.topic())
                    .setTopicId(tp.topicId())
                    .setPartitions(List.of(
                        new FetchResponseData.PartitionData()
                            .setPartitionIndex(tp.topicPartition().partition())
                            .setErrorCode(error.code())
                            .setHighWatermark(hw)
                            .setLastStableOffset(lastStableOffset)
                            .setLogStartOffset(0)
                            .setRecords(records))
                    ))));
    }
```

apalan60 replied:
@FrankYang0529 

Thanks for the review.
I really appreciate your detailed explanation and pointing out my mistake.  
I’ll try out your suggested solution and update the PR accordingly.

apalan60 committed:
Revert "KAFKA-19053: Remove FetchResponse#of which is not used in production anymore"

This reverts commit e50e45e7fadc917f63fc4e0469af1139bb67dfe1.

I reverted this commit because it accidentally deleted a lot of code. Rewriting the refactor from scratch makes it easier to read.

apalan60 committed:
Merge branch 'trunk' into KAFKA-19053-Remove-FetchResponse-of-which-is-not-used-in-production-anymore

apalan60 commented:
> ijuma

@ijuma

Thanks for the review.
I believe I misunderstood the scope of the deletion. I'm currently working on restoring the tests that were mistakenly removed, and trying to construct the FetchResponse without using the ```of``` method, so that the original test logic is preserved while removing the dependency on the ```of``` method.

Thanks again for pointing that out. I'll update the PR soon with an improved version.

apalan60 committed:
Merge branch 'trunk' into KAFKA-19053-Remove-FetchResponse-of-which-is-not-used-in-production-anymore

apalan60 committed:
KAFKA-19053: Remove FetchResponse #of which is not used in production anymore

apalan60 committed:
Merge branch 'trunk' into KAFKA-19053-Remove-FetchResponse-of-which-is-not-used-in-production-anymore

apalan60 commented:
Hi @junrao , @ijuma , @FrankYang0529 
Sorry for the late update.  I’ve now updated the PR with the following adjustments:

In the original version, FetchResponse had two ```#of``` methods: [#of-1](https://github.com/apache/kafka/blob/8c77953d5fa84ce1dfdf83f73560444a4acabc1f/clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java#L224) and [#of-2](https://github.com/apache/kafka/blob/8c77953d5fa84ce1dfdf83f73560444a4acabc1f/clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java#L232).

```#of-1``` was only used in tests and benchmarks, while ```#of-2``` is used by ```FetchSession```.

My current changes only target ```#of-1```. Since it is not used in production, I removed it, and updated the tests that originally called it to instead invoke ```#of-2``` to maintain the original testing logic.

```#of-2``` remains unchanged in this PR because removing it would have a larger impact.

If you see any issues or have any concerns, please let me know and I’ll address them as soon as possible.

ijuma commented:
I don't understand what we are trying to achieve here. It would be good to agree on the goal first. Generally speaking, it's not a strict requirement to remove a method if it's only used by tests. This can make sense if having the overload may lead to bugs in prod due to not passing all the parameters. But then it's helpful to have a test utility method versus adding verbosity to all the tests.

@chia7712 What did you have in mind here?

chia7712 commented:
@ijuma thanks for your explanation. I do agree that there is no strict requirement of removing a test-only method. However, `nodeEndpoints` is crucial for handling specific errors like `NOT_LEADER_OR_FOLLOWER`. Therefore, using this test-only method in production could lead to bugs. We can even add strict check to `toMessage` to ensure the `nodeEndpoints` is not empty when handling the `NOT_LEADER_OR_FOLLOWER` error.

ijuma commented:
OK, if the concern is that this method is unsafe in prod, it's reasonable to replace it with a test utility method.

chia7712 commented:
@apalan60 please fix the conflicts

apalan60 committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-19053-Remove-FetchResponse-of-which-is-not-used-in-production-anymore

apalan60 committed:
Merge remote-tracking branch upstream/trunk into KAFKA-19053-Remove-FetchResponse-of-which-is-not-used-in-production-anymore

apalan60 committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-19053-Remove-FetchResponse-of-which-is-not-used-in-production-anymore

chia7712 reviewed (commented):
@apalan60 Could you please consider using `List.of` instead of `Collections.emptyList()`?

apalan60 committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-19053-Remove-FetchResponse-of-which-is-not-used-in-production-anymore

apalan60 committed:
Address comments

apalan60 commented:
> @apalan60 Could you please consider using `List.of` instead of `Collections.emptyList()`?

Thanks for your review. Updated!

chia7712 commented on clients/src/test/java/org/apache/kafka/clients/FetchSessionHandlerTest.java:
```diff
@@ -217,7 +217,8 @@ public void testSessionless() {
                 respMap(new RespEntry("foo", 0, fooId, 0, 0),
-                        new RespEntry("foo", 1, fooId, 0, 0)));
+                        new RespEntry("foo", 1, fooId, 0, 0)),
+                    Collections.emptyList());
```
could you please use `List.of`?

apalan60 replied:
Thanks for the reminder. Sorry for missing that.

apalan60 committed:
Address comments

apalan60 committed:
chore: fix style

apalan60 committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-19053-Remove-FetchResponse-of-which-is-not-used-in-production-anymore
